File: backend/src/models/book.model.js
const mongoose = require('mongoose');

const bookSchema = new mongoose.Schema({
  title: {
    type: String,
    required: [true, 'A book must have a title'],
    trim: true,
    maxlength: [100, 'Title cannot exceed 100 characters']
  },
  author: {
    type: String,
    required: [true, 'A book must have an author'],
    trim: true,
    maxlength: [50, 'Author name cannot exceed 50 characters']
  },
  description: {
    type: String,
    trim: true,
    maxlength: [2000, 'Description cannot exceed 2000 characters']
  },
  imageUrl: {
    type: String,
    default: 'default-book.jpg'
  },
  condition: {
    type: String,
    enum: ['New', 'Like New', 'Good', 'Fair', 'Poor'],
    default: 'Good'
  },
  price: {
    type: Number,
    default: 0
  },
  category: {
    type: String,
    required: [true, 'A book must have a category'],
    trim: true
  },
  exchangeType: {
    type: String,
    enum: ['Sell', 'Exchange', 'Free', 'Rent'],
    default: 'Sell'
  },
  department: {
    type: String,
    trim: true
  },
  courseCode: {
    type: String,
    trim: true
  },
  owner: {
    type: mongoose.Schema.ObjectId,
    ref: 'User',
    required: [true, 'A book must belong to an owner']
  },
  isbn: {
    type: String,
    trim: true
  },
  language: {
    type: String,
    default: 'English'
  },
  publisher: {
    type: String,
    trim: true
  },
  publishedYear: {
    type: Number
  },
  availability: {
    type: Boolean,
    default: true
  },
  createdAt: {
    type: Date,
    default: Date.now()
  },
  ratings: [{
    user: {
      type: mongoose.Schema.ObjectId,
      ref: 'User'
    },
    rating: {
      type: Number,
      min: 1,
      max: 5
    },
    review: String,
    createdAt: {
      type: Date,
      default: Date.now
    }
  }]
}, {
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Calculate average rating
bookSchema.virtual('averageRating').get(function() {
  if (this.ratings.length === 0) return 0;
  
  const sum = this.ratings.reduce((total, rating) => total + rating.rating, 0);
  return (sum / this.ratings.length).toFixed(1);
});

// Index for searching
bookSchema.index({ title: 'text', author: 'text', description: 'text' });

module.exports = mongoose.model('Book', bookSchema); 

File: lib/models/book_model.dart
class Book {
  final String id;
  final String title;
  final String author;
  final String? description;
  final String? imageUrl;
  final String condition;
  final double price;
  final String category;
  final String exchangeType;
  final String? department;
  final String? courseCode;
  final Map<String, dynamic> owner;
  final String? isbn;
  final String language;
  final String? publisher;
  final int? publishedYear;
  final bool availability;
  final DateTime createdAt;
  final List<dynamic>? ratings;
  final double averageRating;

  Book({
    required this.id,
    required this.title,
    required this.author,
    this.description,
    this.imageUrl,
    required this.condition,
    required this.price,
    required this.category,
    required this.exchangeType,
    this.department,
    this.courseCode,
    required this.owner,
    this.isbn,
    required this.language,
    this.publisher,
    this.publishedYear,
    required this.availability,
    required this.createdAt,
    this.ratings,
    this.averageRating = 0.0,
  });

  factory Book.fromJson(Map<String, dynamic> json) {
    return Book(
      id: json['_id'],
      title: json['title'],
      author: json['author'],
      description: json['description'],
      imageUrl: json['imageUrl'],
      condition: json['condition'],
      price:
          json['price'] != null ? double.parse(json['price'].toString()) : 0.0,
      category: json['category'],
      exchangeType: json['exchangeType'],
      department: json['department'],
      courseCode: json['courseCode'],
      owner: json['owner'],
      isbn: json['isbn'],
      language: json['language'] ?? 'English',
      publisher: json['publisher'],
      publishedYear: json['publishedYear'],
      availability: json['availability'] ?? true,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'])
          : DateTime.now(),
      ratings: json['ratings'],
      averageRating: json['averageRating'] != null
          ? double.parse(json['averageRating'].toString())
          : 0.0,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'title': title,
      'author': author,
      'description': description,
      'imageUrl': imageUrl,
      'condition': condition,
      'price': price,
      'category': category,
      'exchangeType': exchangeType,
      'department': department,
      'courseCode': courseCode,
      'owner': owner,
      'isbn': isbn,
      'language': language,
      'publisher': publisher,
      'publishedYear': publishedYear,
      'availability': availability,
      'createdAt': createdAt.toIso8601String(),
      'ratings': ratings,
      'averageRating': averageRating,
    };
  }
}
